File: 8.回溯/21.解数独/37.解数独1.py
#
# @lc app=leetcode.cn id=37 lang=python3
#
# [37] 解数独
#
from mytools import *
# @lc code=start
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        empty = []
        row_s = [set() for _ in range(9)]
        col_s = [set() for _ in range(9)]
        box_s = [set() for _ in range(9)]
        m, n = len(board), len(board[0])
        for i in range(m):
            for j in range(n):
                if board[i][j] == '.':
                    empty.append((i, j))
                else:
                    row_s[i].add(board[i][j])
                    col_s[j].add(board[i][j])
                    box_idx = i // 3 * 3 + j // 3
                    box_s[box_idx].add(board[i][j])


        def dfs(pos):
            if (pos == len(empty)): return True
            x, y = empty[pos]
            bi = x // 3 * 3 + y // 3
            for i in range(1, 10):
                if str(i) in row_s[x] or str(i) in col_s[y] or str(i) in box_s[bi]: continue
                row_s[x].add(str(i))
                col_s[y].add(str(i))
                box_s[bi].add(str(i))
                board[x][y] = str(i)
                if dfs(pos + 1) : return True
                board[x][y] = '.'
                row_s[x].remove(str(i))
                col_s[y].remove(str(i))
                box_s[bi].remove(str(i))
            return False
        
        return dfs(0)



# @lc code=end

File: 8.回溯/21.解数独/37.解数独1.cpp
/*
 * @lc app=leetcode.cn id=37 lang=cpp
 *
 * [37] 解数独
 */
#include "tools.h"
// @lc code=start
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
        vector<vector<int>> row_vis(9, vector<int>(10, 0));
        vector<vector<int>> col_vis(9, vector<int>(10, 0));
        vector<vector<int>> box_vis(9, vector<int>(10, 0));
        vector<pair<int, int>> _e;
        for (int i = 0; i < 9; i ++ )
        {
            for (int j = 0; j < 9; j ++ )
            {
                int box_idx = i / 3 * 3 + j / 3;
                if (board[i][j] != '.')
                {
                    int _x = board[i][j] - '0';
                    row_vis[i][_x] = 1;
                    col_vis[j][_x] = 1;
                    box_vis[box_idx][_x] = 1;
                }
                else
                {
                    _e.emplace_back(i, j);
                }
            }
        }
        auto dfs = [&](auto&& self, int pos) -> bool
        {
            if (pos == _e.size()) return 1;
            auto d = _e[pos];
            int x = d.first, y = d.second;
            int bi = x / 3 * 3 + y / 3;
            for (int i = 1; i < 10; i ++ )
            {
                if (!row_vis[x][i] && !col_vis[y][i] && !box_vis[bi][i])
                {
                    row_vis[x][i] = 1, col_vis[y][i] = 1, box_vis[bi][i] = 1;
                    board[x][y] = i + '0';
                    if (self(self, pos + 1)) return 1;
                    board[x][y] = '.';
                    row_vis[x][i] = 0, col_vis[y][i] = 0, box_vis[bi][i] = 0;
                }
            }
            return 0;
        };
        bool tmp = dfs(dfs, 0);
    }
};
// @lc code=end

